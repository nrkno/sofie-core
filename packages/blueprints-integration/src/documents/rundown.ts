import type { UserEditingDefinition } from '../userEditing'
import type { RundownPlaylistTiming } from './playlistTiming'

/** The Rundown generated from Blueprint */

export interface IBlueprintRundown<TPrivateData = unknown, TPublicData = unknown> {
	externalId: string
	/** Rundown slug - user-presentable name */
	name: string

	/** Rundown description: Longer user-presentable description of the rundown */
	description?: string

	/** Rundown timing information */
	timing: RundownPlaylistTiming

	/** Arbitraty data storage for internal use in the blueprints */
	privateData?: TPrivateData
	/** Arbitraty data relevant for other systems, made available to them through APIs */
	publicData?: TPublicData

	/** A hint to the Core that the Rundown should be a part of a playlist */
	playlistExternalId?: string

	/**
	 * Whether the end of the rundown marks a break in the show.
	 * Allows the Next Break timer in the Rundown Header to time to the end of this rundown when looking for the next break.
	 */
	endOfRundownIsShowBreak?: boolean

	/**
	 * User editing definitions for this rundown
	 */
	userEditOperations?: UserEditingDefinition[]
}
/** The Rundown sent from Core */

export interface IBlueprintRundownDB<TPrivateData = unknown, TPublicData = unknown>
	extends IBlueprintRundown<TPrivateData, TPublicData>,
		IBlueprintRundownDBData {}
/** Properties added to a rundown in Core */

export interface IBlueprintRundownDBData {
	_id: string

	/** Id of the showStyle variant used */
	showStyleVariantId: string

	/** RundownPlaylist this rundown is member of */
	playlistId?: string

	/** Air-status, comes from NCS, examples: "READY" | "NOT READY" */
	airStatus?: string
}

export interface IBlueprintSegmentRundown<TPrivateData = unknown, TPublicData = unknown> {
	externalId: string

	/** Arbitraty data storage for internal use in the blueprints */
	privateData?: TPrivateData
	/** Arbitraty data relevant for other systems, made available to them through APIs */
	publicData?: TPublicData
}
