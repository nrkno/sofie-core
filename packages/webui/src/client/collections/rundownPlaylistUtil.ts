import { RundownId, RundownPlaylistId } from '@sofie-automation/corelib/dist/dataModel/Ids'
import { Rundown, DBRundown } from '@sofie-automation/corelib/dist/dataModel/Rundown'
import { DBRundownPlaylist } from '@sofie-automation/corelib/dist/dataModel/RundownPlaylist'
import { normalizeArrayToMap } from '@sofie-automation/corelib/dist/lib'
import { sortRundownIDsInPlaylist } from '@sofie-automation/corelib/dist/playout/playlist'
import _ from 'underscore'
import { Rundowns } from './index'
import { FindOptions } from './lib'
import { MongoQuery } from '@sofie-automation/corelib/dist/mongo'

/**
 * Direct database accessors for the RundownPlaylist
 * These used to reside on the Rundown class
 */
export class RundownPlaylistCollectionUtil {
	/** Returns an array of all Rundowns in the RundownPlaylist, sorted in playout order */
	static getRundownsOrdered(
		playlist: Pick<DBRundownPlaylist, '_id' | 'rundownIdsInOrder'>,
		selector?: MongoQuery<Rundown>,
		options?: FindOptions<Rundown>
	): Rundown[] {
		const allRundowns = RundownPlaylistCollectionUtil.getRundownsUnordered(playlist._id, selector, options)

		const rundownsMap = normalizeArrayToMap(allRundowns, '_id')

		const sortedIds = sortRundownIDsInPlaylist(playlist.rundownIdsInOrder, Array.from(rundownsMap.keys()))

		return _.compact(sortedIds.map((id) => rundownsMap.get(id)))
	}
	/** Returns an array of all Rundowns in the RundownPlaylist, in no predictable order */
	static getRundownsUnordered(
		playlistId: RundownPlaylistId,
		selector?: MongoQuery<Rundown>,
		options?: FindOptions<Rundown>
	): Rundown[] {
		return Rundowns.find(
			{
				playlistId: playlistId,
				...selector,
			},
			{
				sort: {
					_id: 1,
				},
				...options,
			}
		).fetch()
	}
	/** Returns an array with the id:s of all Rundowns in the RundownPlaylist, sorted in playout order */
	static getRundownOrderedIDs(playlist: Pick<DBRundownPlaylist, '_id' | 'rundownIdsInOrder'>): RundownId[] {
		const allIds = RundownPlaylistCollectionUtil.getRundownUnorderedIDs(playlist)

		return sortRundownIDsInPlaylist(playlist.rundownIdsInOrder, allIds)
	}
	/** Returns an array with the id:s of all Rundowns in the RundownPlaylist, in no predictable order */
	static getRundownUnorderedIDs(playlist: Pick<DBRundownPlaylist, '_id'>): RundownId[] {
		const rundowns = Rundowns.find(
			{
				playlistId: playlist._id,
			},
			{
				fields: {
					_id: 1,
				},
			}
		).fetch() as Array<Pick<DBRundown, '_id'>>

		return rundowns.map((i) => i._id)
	}
}
