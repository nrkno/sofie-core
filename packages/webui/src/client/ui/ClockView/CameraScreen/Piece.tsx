import React, { useContext, useMemo } from 'react'
import { PartId } from '@sofie-automation/corelib/dist/dataModel/Ids'
import classNames from 'classnames'
import { CanvasSizeContext } from '.'
import { PieceExtended } from '../../../lib/RundownResolver'
import { PieceElement } from '../../SegmentContainer/PieceElement'
import { getSplitItems } from '../../SegmentContainer/getSplitItems'

const PIECE_TYPE_INDICATOR_BORDER_RADIUS = 11

export const Piece = React.memo(function Piece({
	piece,
	partId,
	left,
	width,
	zoom,
	isLive,
}: {
	piece: PieceExtended
	partId: PartId
	left: number
	width: number | null
	zoom: number
	isLive: boolean
}): JSX.Element | null {
	const canvasWidth = useContext(CanvasSizeContext)

	const indicatorPadding = -1 * PIECE_TYPE_INDICATOR_BORDER_RADIUS
	let pixelLeft = Math.max(indicatorPadding, left * zoom) + PIECE_TYPE_INDICATOR_BORDER_RADIUS
	let pixelWidth = Math.min(canvasWidth, (width ?? canvasWidth) * zoom + Math.min(left * zoom, 0))

	if (isLive) {
		pixelLeft = 0
		pixelWidth = Math.min(canvasWidth, (width ?? canvasWidth) * zoom) + PIECE_TYPE_INDICATOR_BORDER_RADIUS
	}

	const style = useMemo<React.CSSProperties>(
		() =>
			width === null
				? {
						width: `100%`,
						transform: `translate(${pixelLeft}px, 0)`,
				  }
				: {
						width: `${pixelWidth}px`,
						transform: `translate(${pixelLeft}px, 0)`,
				  },
		[width, pixelLeft, pixelWidth]
	)

	return (
		<div className={classNames('camera-screen__piece', { live: isLive })}>
			{pixelLeft < canvasWidth ? (
				<PieceElement
					className="camera-screen__piece-background"
					layer={piece.sourceLayer}
					piece={piece}
					partId={partId}
					style={style}
				>
					{getSplitItems(piece, 'camera-screen__piece-sub-background')}
				</PieceElement>
			) : null}
			<PieceElement
				className="camera-screen__piece-type-indicator"
				partId={partId}
				layer={piece.sourceLayer}
				piece={piece}
			>
				{getSplitItems(piece, 'camera-screen__piece-type-indicator-sub-background')}
			</PieceElement>
			<div className="camera-screen__piece-label">{piece.instance.piece.name}</div>
		</div>
	)
})
