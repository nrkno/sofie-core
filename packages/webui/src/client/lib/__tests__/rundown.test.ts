import {
	setupDefaultStudioEnvironment,
	DefaultEnvironment,
	setupDefaultRundownPlaylist,
	convertToUIShowStyleBase,
} from '../../../__mocks__/helpers/database.js'
import { RundownUtils } from '../rundown.js'
import { Piece } from '@sofie-automation/corelib/dist/dataModel/Piece'
import { defaultPartInstance, defaultPiece, defaultPieceInstance } from '../../../__mocks__/defaultCollectionObjects.js'
import { protectString, unprotectString } from '@sofie-automation/corelib/dist/protectedString'
import { PieceLifespan } from '@sofie-automation/blueprints-integration'
import { PartInstance } from '@sofie-automation/meteor-lib/dist/collections/PartInstances'
import { PieceInstance } from '@sofie-automation/corelib/dist/dataModel/PieceInstance'
import { RundownPlaylistId } from '@sofie-automation/corelib/dist/dataModel/Ids'
import { PartInstances, PieceInstances, Pieces, RundownPlaylists } from '../../collections/index.js'
import { MongoMock } from '../../../__mocks__/mongo.js'
import { RundownPlaylistCollectionUtil } from '../../collections/rundownPlaylistUtil.js'
import { RundownPlaylistClientUtil } from '../rundownPlaylistUtil.js'

const mockRundownPlaylistsCollection = MongoMock.getInnerMockCollection(RundownPlaylists)
const mockPartInstancesCollection = MongoMock.getInnerMockCollection(PartInstances)
const mockPieceInstancesCollection = MongoMock.getInnerMockCollection(PieceInstances)
const mockPiecesCollection = MongoMock.getInnerMockCollection(Pieces)

// This is a hack, the tests should be rewriten to not use methods unrelated to the testee
jest.mock('../../ui/Collections', () => {
	const mockClientCollections = jest.requireActual('../../ui/Collections')
	const mockLibCollections = jest.requireActual('../../collections/index')
	return {
		...mockClientCollections,
		UIParts: mockLibCollections.Parts, // for most purposes they're equivalent
		UIPartInstances: mockLibCollections.PartInstances, // for most purposes they're equivalent
	}
})

describe('client/lib/rundown', () => {
	let env: DefaultEnvironment
	let playlistId: RundownPlaylistId
	beforeEach(async () => {
		env = await setupDefaultStudioEnvironment()
		playlistId = (await setupDefaultRundownPlaylist(env)).playlistId
	})
	describe('RundownUtils.getResolvedSegment', () => {
		test('Basic Segment resolution', () => {
			const showStyleBase = convertToUIShowStyleBase(env.showStyleBase)
			const playlist = RundownPlaylists.findOne(playlistId)
			if (!playlist) throw new Error('Rundown not found')

			const { currentPartInstance, nextPartInstance } =
				RundownPlaylistClientUtil.getSelectedPartInstances(playlist)

			const rundowns = RundownPlaylistCollectionUtil.getRundownsOrdered(playlist)
			const { parts, segments } = RundownPlaylistClientUtil.getSegmentsAndPartsSync(playlist)
			const rundown = rundowns[0]
			const segment = segments[0]

			const resolvedSegment = RundownUtils.getResolvedSegment(
				showStyleBase,
				playlist,
				rundown,
				segment,
				new Set(segments.slice(0, 0).map((s) => s._id)),
				[],
				new Map(),
				parts.map((part) => part._id),
				currentPartInstance,
				nextPartInstance
			)
			expect(resolvedSegment).toBeTruthy()
			expect(resolvedSegment.parts).toHaveLength(2)
			expect(resolvedSegment).toMatchObject({
				isLiveSegment: false,
				isNextSegment: false,
				currentLivePart: undefined,
				currentNextPart: undefined,
				hasRemoteItems: false,
				hasGuestItems: false,
				hasAlreadyPlayed: false,
				autoNextPart: false,
			})
			expect(resolvedSegment).toMatchSnapshot()
		})

		test('Infinite Piece starting in first Part is populated across the Segment', () => {
			const showStyleBase = convertToUIShowStyleBase(env.showStyleBase)
			const sourceLayerIds = Object.keys(showStyleBase.sourceLayers)
			const outputLayerIds = Object.keys(showStyleBase.outputLayers)

			const playlist = RundownPlaylists.findOne(playlistId)
			if (!playlist) throw new Error('Playlist not found')

			const { currentPartInstance, nextPartInstance } =
				RundownPlaylistClientUtil.getSelectedPartInstances(playlist)

			const rundowns = RundownPlaylistCollectionUtil.getRundownsOrdered(playlist)
			const { parts, segments } = RundownPlaylistClientUtil.getSegmentsAndPartsSync(playlist)
			const rundown = rundowns[0]
			const rundownId = rundown._id
			const segment = segments[1]
			const firstPart = parts.find((part) => part.segmentId === segment._id)

			if (!firstPart) throw new Error('Mock Segment 1 not found')

			const infinitePiece: Piece = {
				...defaultPiece(protectString(rundownId + '_infinite_piece'), rundownId, segment._id, firstPart._id),
				externalId: 'MOCK_INFINITE_PIECE',
				name: 'Infinite',
				sourceLayerId: sourceLayerIds[0],
				outputLayerId: outputLayerIds[0],
				enable: {
					start: 1000,
				},
				lifespan: PieceLifespan.OutOnSegmentEnd,
			}

			mockPiecesCollection.insert(infinitePiece)

			const resolvedSegment = RundownUtils.getResolvedSegment(
				showStyleBase,
				playlist,
				rundown,
				segment,
				new Set(segments.slice(0, 1).map((s) => s._id)),
				[],
				new Map(),
				parts.map((part) => part._id),
				currentPartInstance,
				nextPartInstance
			)
			expect(resolvedSegment).toBeTruthy()
			expect(resolvedSegment.parts).toHaveLength(3)
			expect(resolvedSegment).toMatchObject({
				isLiveSegment: false,
				isNextSegment: false,
				currentLivePart: undefined,
				currentNextPart: undefined,
				hasRemoteItems: false,
				hasGuestItems: false,
				hasAlreadyPlayed: false,
				autoNextPart: false,
			})
			const resolvedInfinitePiece00 = resolvedSegment.parts[0].pieces.find(
				(piece) => piece.instance.piece._id === infinitePiece._id
			)
			expect(resolvedInfinitePiece00).toBeDefined()
			expect(resolvedInfinitePiece00?.renderedInPoint).toBe(1000)
			expect(resolvedInfinitePiece00?.renderedDuration).toBe(null)
			const resolvedInfinitePiece01 = resolvedSegment.parts[1].pieces.find(
				(piece) => piece.instance.piece._id === infinitePiece._id
			)
			expect(resolvedInfinitePiece01).toBeDefined()
			expect(resolvedInfinitePiece01?.renderedInPoint).toBe(0)
			expect(resolvedInfinitePiece01?.renderedDuration).toBe(null)
		})

		test('Infinite Piece starting in first Part is cropped by another Piece', () => {
			const showStyleBase = convertToUIShowStyleBase(env.showStyleBase)
			const sourceLayerIds = Object.keys(showStyleBase.sourceLayers)
			const outputLayerIds = Object.keys(showStyleBase.outputLayers)

			const playlist = RundownPlaylists.findOne(playlistId)
			if (!playlist) throw new Error('Playlist not found')

			const { currentPartInstance, nextPartInstance } =
				RundownPlaylistClientUtil.getSelectedPartInstances(playlist)

			const rundowns = RundownPlaylistCollectionUtil.getRundownsOrdered(playlist)
			const { parts, segments } = RundownPlaylistClientUtil.getSegmentsAndPartsSync(playlist)
			const rundown = rundowns[0]
			const rundownId = rundown._id
			const segment = segments[1]
			const firstPart = parts.find((part) => part.segmentId === segment._id)

			if (!firstPart) throw new Error('Mock Segment 1 not found')

			const infinitePiece: Piece = {
				...defaultPiece(protectString(rundownId + '_infinite_piece'), rundownId, segment._id, firstPart._id),
				externalId: 'MOCK_INFINITE_PIECE',
				name: 'Infinite',
				sourceLayerId: sourceLayerIds[0],
				outputLayerId: outputLayerIds[0],
				enable: {
					start: 1000,
				},
				lifespan: PieceLifespan.OutOnSegmentChange,
			}
			mockPiecesCollection.insert(infinitePiece)

			const croppingPiece: Piece = {
				...defaultPiece(protectString(rundownId + '_cropping_piece'), rundownId, segment._id, firstPart._id),
				externalId: 'MOCK_CROPPING_PIECE',
				name: 'Cropping',
				sourceLayerId: sourceLayerIds[0],
				outputLayerId: outputLayerIds[0],
				enable: {
					start: 4000,
					duration: 1000,
				},
				lifespan: PieceLifespan.WithinPart,
			}
			mockPiecesCollection.insert(croppingPiece)

			const resolvedSegment = RundownUtils.getResolvedSegment(
				showStyleBase,
				playlist,
				rundown,
				segment,
				new Set(segments.slice(0, 1).map((s) => s._id)),
				[],
				new Map(),
				parts.map((part) => part._id),
				currentPartInstance,
				nextPartInstance
			)
			expect(resolvedSegment).toBeTruthy()
			expect(resolvedSegment.parts).toHaveLength(3)
			expect(resolvedSegment).toMatchObject({
				isLiveSegment: false,
				isNextSegment: false,
				currentLivePart: undefined,
				currentNextPart: undefined,
				hasRemoteItems: false,
				hasGuestItems: false,
				hasAlreadyPlayed: false,
				autoNextPart: false,
			})

			expect(resolvedSegment.parts[0].pieces).toHaveLength(2)

			const resolvedInfinitePiece00 = resolvedSegment.parts[0].pieces.find(
				(piece) => piece.instance.piece._id === infinitePiece._id
			)
			expect(resolvedInfinitePiece00).toBeDefined()
			expect(resolvedInfinitePiece00?.renderedInPoint).toBe(1000)
			expect(resolvedInfinitePiece00?.renderedDuration).toBe(3000)

			const resolvedCroppingPiece00 = resolvedSegment.parts[0].pieces.find(
				(piece) => piece.instance.piece._id === croppingPiece._id
			)
			expect(resolvedCroppingPiece00).toBeDefined()
			expect(resolvedCroppingPiece00?.renderedInPoint).toBe(4000)
			expect(resolvedCroppingPiece00?.renderedDuration).toBe(1000)

			expect(resolvedSegment.parts[1].pieces).toHaveLength(0)
		})

		test("User-Stopped Infinite Piece starting in first Part maintains it's length when followed by another Piece", () => {
			const showStyleBase = convertToUIShowStyleBase(env.showStyleBase)
			const sourceLayerIds = Object.keys(showStyleBase.sourceLayers)
			const outputLayerIds = Object.keys(showStyleBase.outputLayers)

			const playlistActivationId = protectString('mock_activation_0')
			mockRundownPlaylistsCollection.update(unprotectString(playlistId), {
				$set: {
					activationId: playlistActivationId,
				},
			})

			let playlist = RundownPlaylists.findOne(playlistId)
			if (!playlist) throw new Error('Playlist not found')

			const rundowns = RundownPlaylistCollectionUtil.getRundownsOrdered(playlist)
			const { parts, segments } = RundownPlaylistClientUtil.getSegmentsAndPartsSync(playlist)
			const rundown = rundowns[0]
			const rundownId = rundown._id
			const segment = segments[1]
			const firstPart = parts.find((part) => part.segmentId === segment._id)

			if (!firstPart) throw new Error('Mock Segment 1 not found')

			const infinitePiece: Piece = {
				...defaultPiece(protectString(rundownId + '_infinite_piece'), rundownId, segment._id, firstPart._id),
				externalId: 'MOCK_INFINITE_PIECE',
				name: 'Infinite',
				sourceLayerId: sourceLayerIds[0],
				outputLayerId: outputLayerIds[0],
				enable: {
					start: 1000,
				},
				lifespan: PieceLifespan.OutOnSegmentChange,
			}
			mockPiecesCollection.insert(infinitePiece)

			const followingPiece: Piece = {
				...defaultPiece(protectString(rundownId + '_cropping_piece'), rundownId, segment._id, firstPart._id),
				externalId: 'MOCK_CROPPING_PIECE',
				name: 'Cropping',
				sourceLayerId: sourceLayerIds[0],
				outputLayerId: outputLayerIds[0],
				enable: {
					start: 4000,
					duration: 1000,
				},
				lifespan: PieceLifespan.WithinPart,
			}
			mockPiecesCollection.insert(followingPiece)

			const segmentPlayoutId = protectString('mock_segment_playout_0')
			const mockCurrentPartInstance: PartInstance = {
				...defaultPartInstance(
					protectString(rundownId + '_partInstance_0'),
					playlistActivationId,
					segmentPlayoutId,
					firstPart
				),
			}

			mockPartInstancesCollection.insert(mockCurrentPartInstance)

			const infinitePieceInstance: PieceInstance = {
				...defaultPieceInstance(
					protectString('instance_' + infinitePiece._id),
					playlistActivationId,
					rundown._id,
					mockCurrentPartInstance._id,
					infinitePiece
				),
				userDuration: {
					endRelativeToPart: 2000,
				},
			}

			mockPieceInstancesCollection.insert(infinitePieceInstance)

			const followingPieceInstance: PieceInstance = {
				...defaultPieceInstance(
					protectString('instance_' + followingPiece._id),
					playlistActivationId,
					rundown._id,
					mockCurrentPartInstance._id,
					followingPiece
				),
			}

			mockPieceInstancesCollection.insert(followingPieceInstance)

			mockRundownPlaylistsCollection.update(unprotectString(playlistId), {
				$set: {
					currentPartInfo: {
						partInstanceId: mockCurrentPartInstance._id,
						rundownId: mockCurrentPartInstance.rundownId,
						manuallySelected: false,
						consumesQueuedSegmentId: false,
					},
				},
			})

			playlist = RundownPlaylists.findOne(playlistId)
			if (!playlist) throw new Error('Playlist not found')
			const { currentPartInstance, nextPartInstance } =
				RundownPlaylistClientUtil.getSelectedPartInstances(playlist)

			const resolvedSegment = RundownUtils.getResolvedSegment(
				showStyleBase,
				playlist,
				rundown,
				segment,
				new Set(segments.slice(0, 1).map((s) => s._id)),
				[],
				new Map(),
				parts.map((part) => part._id),
				currentPartInstance,
				nextPartInstance
			)
			expect(resolvedSegment).toBeTruthy()
			expect(resolvedSegment.parts).toHaveLength(3)
			expect(resolvedSegment).toMatchObject({
				isLiveSegment: true,
				isNextSegment: false,
				currentNextPart: undefined,
				hasRemoteItems: false,
				hasGuestItems: false,
				autoNextPart: false,
			})

			expect(resolvedSegment.parts[0].pieces).toHaveLength(2)

			const resolvedInfinitePiece00 = resolvedSegment.parts[0].pieces.find(
				(piece) => piece.instance._id === infinitePieceInstance._id
			)
			expect(resolvedInfinitePiece00).toBeDefined()
			expect(resolvedInfinitePiece00?.renderedInPoint).toBe(1000)
			expect(resolvedInfinitePiece00?.renderedDuration).toBe(1000)

			const resolvedCroppingPiece00 = resolvedSegment.parts[0].pieces.find(
				(piece) => piece.instance._id === followingPieceInstance._id
			)
			expect(resolvedCroppingPiece00).toBeDefined()
			expect(resolvedCroppingPiece00?.renderedInPoint).toBe(4000)
			expect(resolvedCroppingPiece00?.renderedDuration).toBe(1000)

			expect(resolvedSegment.parts[1].pieces).toHaveLength(0)
		})
	})
})
