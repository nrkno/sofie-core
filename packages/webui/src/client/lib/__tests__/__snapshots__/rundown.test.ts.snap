// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`client/lib/rundown RundownUtils.getResolvedSegment Basic Segment resolution 1`] = `
{
  "autoNextPart": false,
  "currentLivePart": undefined,
  "currentNextPart": undefined,
  "hasAlreadyPlayed": false,
  "hasGuestItems": false,
  "hasRemoteItems": false,
  "isLiveSegment": false,
  "isNextSegment": false,
  "parts": [
    {
      "instance": {
        "_id": "randomId9002_part0_0_tmp_instance",
        "isTemporary": true,
        "part": {
          "_id": "randomId9002_part0_0",
          "_rank": 0,
          "externalId": "MOCK_PART_0_0",
          "rundownId": "randomId9002",
          "segmentId": "randomId9002_segment0",
          "title": "Part 0 0",
        },
        "playlistActivationId": "",
        "rehearsal": false,
        "rundownId": "randomId9002",
        "segmentId": "randomId9002_segment0",
        "segmentPlayoutId": "",
        "takeCount": -1,
      },
      "partId": "randomId9002_part0_0",
      "pieces": [
        {
          "instance": {
            "_id": "randomId9002_part0_0_tmp_instance_randomId9002_piece001",
            "isTemporary": true,
            "partInstanceId": "randomId9002_part0_0_tmp_instance",
            "piece": {
              "_id": "randomId9002_piece001",
              "content": {},
              "enable": {
                "start": 0,
              },
              "externalId": "MOCK_PIECE_001",
              "invalid": false,
              "lifespan": "part-only",
              "name": "Piece 001",
              "outputLayerId": "pgm",
              "pieceType": "normal",
              "sourceLayerId": "vt0",
              "startPartId": "randomId9002_part0_0",
              "startRundownId": "randomId9002",
              "startSegmentId": "randomId9002_segment0",
              "timelineObjectsString": "[]",
            },
            "playlistActivationId": "",
            "priority": 5,
            "rundownId": "randomId9002",
          },
          "outputLayer": {
            "_id": "pgm",
            "_rank": 0,
            "isPGM": true,
            "name": "PGM",
            "sourceLayers": [
              {
                "_id": "vt0",
                "_rank": 1,
                "exclusiveGroup": "main",
                "followingItems": [],
                "name": "VT",
                "pieces": [
                  [Circular],
                ],
                "type": 2,
              },
              {
                "_id": "cam0",
                "_rank": 0,
                "exclusiveGroup": "main",
                "followingItems": [],
                "name": "Camera",
                "pieces": [
                  {
                    "instance": {
                      "_id": "randomId9002_part0_1_tmp_instance_randomId9002_piece010",
                      "isTemporary": true,
                      "partInstanceId": "randomId9002_part0_1_tmp_instance",
                      "piece": {
                        "_id": "randomId9002_piece010",
                        "content": {},
                        "enable": {
                          "start": 0,
                        },
                        "externalId": "MOCK_PIECE_010",
                        "invalid": false,
                        "lifespan": "part-only",
                        "name": "Piece 010",
                        "outputLayerId": "pgm",
                        "pieceType": "normal",
                        "sourceLayerId": "cam0",
                        "startPartId": "randomId9002_part0_1",
                        "startRundownId": "randomId9002",
                        "startSegmentId": "randomId9002_segment0",
                        "timelineObjectsString": "[]",
                      },
                      "playlistActivationId": "",
                      "priority": 5,
                      "rundownId": "randomId9002",
                    },
                    "outputLayer": [Circular],
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "sourceLayer": [Circular],
                  },
                ],
                "type": 1,
              },
            ],
            "used": true,
          },
          "renderedDuration": null,
          "renderedInPoint": 0,
          "sourceLayer": {
            "_id": "vt0",
            "_rank": 1,
            "exclusiveGroup": "main",
            "followingItems": [],
            "name": "VT",
            "pieces": [
              [Circular],
            ],
            "type": 2,
          },
        },
      ],
      "renderedDuration": 3000,
      "startsAt": 0,
      "willProbablyAutoNext": false,
    },
    {
      "instance": {
        "_id": "randomId9002_part0_1_tmp_instance",
        "isTemporary": true,
        "part": {
          "_id": "randomId9002_part0_1",
          "_rank": 1,
          "externalId": "MOCK_PART_0_1",
          "rundownId": "randomId9002",
          "segmentId": "randomId9002_segment0",
          "title": "Part 0 1",
        },
        "playlistActivationId": "",
        "rehearsal": false,
        "rundownId": "randomId9002",
        "segmentId": "randomId9002_segment0",
        "segmentPlayoutId": "",
        "takeCount": -1,
      },
      "partId": "randomId9002_part0_1",
      "pieces": [
        {
          "instance": {
            "_id": "randomId9002_part0_1_tmp_instance_randomId9002_piece010",
            "isTemporary": true,
            "partInstanceId": "randomId9002_part0_1_tmp_instance",
            "piece": {
              "_id": "randomId9002_piece010",
              "content": {},
              "enable": {
                "start": 0,
              },
              "externalId": "MOCK_PIECE_010",
              "invalid": false,
              "lifespan": "part-only",
              "name": "Piece 010",
              "outputLayerId": "pgm",
              "pieceType": "normal",
              "sourceLayerId": "cam0",
              "startPartId": "randomId9002_part0_1",
              "startRundownId": "randomId9002",
              "startSegmentId": "randomId9002_segment0",
              "timelineObjectsString": "[]",
            },
            "playlistActivationId": "",
            "priority": 5,
            "rundownId": "randomId9002",
          },
          "outputLayer": {
            "_id": "pgm",
            "_rank": 0,
            "isPGM": true,
            "name": "PGM",
            "sourceLayers": [
              {
                "_id": "vt0",
                "_rank": 1,
                "exclusiveGroup": "main",
                "followingItems": [],
                "name": "VT",
                "pieces": [
                  {
                    "instance": {
                      "_id": "randomId9002_part0_0_tmp_instance_randomId9002_piece001",
                      "isTemporary": true,
                      "partInstanceId": "randomId9002_part0_0_tmp_instance",
                      "piece": {
                        "_id": "randomId9002_piece001",
                        "content": {},
                        "enable": {
                          "start": 0,
                        },
                        "externalId": "MOCK_PIECE_001",
                        "invalid": false,
                        "lifespan": "part-only",
                        "name": "Piece 001",
                        "outputLayerId": "pgm",
                        "pieceType": "normal",
                        "sourceLayerId": "vt0",
                        "startPartId": "randomId9002_part0_0",
                        "startRundownId": "randomId9002",
                        "startSegmentId": "randomId9002_segment0",
                        "timelineObjectsString": "[]",
                      },
                      "playlistActivationId": "",
                      "priority": 5,
                      "rundownId": "randomId9002",
                    },
                    "outputLayer": [Circular],
                    "renderedDuration": null,
                    "renderedInPoint": 0,
                    "sourceLayer": [Circular],
                  },
                ],
                "type": 2,
              },
              {
                "_id": "cam0",
                "_rank": 0,
                "exclusiveGroup": "main",
                "followingItems": [],
                "name": "Camera",
                "pieces": [
                  [Circular],
                ],
                "type": 1,
              },
            ],
            "used": true,
          },
          "renderedDuration": null,
          "renderedInPoint": 0,
          "sourceLayer": {
            "_id": "cam0",
            "_rank": 0,
            "exclusiveGroup": "main",
            "followingItems": [],
            "name": "Camera",
            "pieces": [
              [Circular],
            ],
            "type": 1,
          },
        },
      ],
      "renderedDuration": 3000,
      "startsAt": 3000,
      "willProbablyAutoNext": false,
    },
  ],
  "segmentExtended": {
    "_id": "randomId9002_segment0",
    "_rank": 0,
    "externalId": "MOCK_SEGMENT_0",
    "name": "Segment 0",
    "outputLayers": {
      "pgm": {
        "_id": "pgm",
        "_rank": 0,
        "isPGM": true,
        "name": "PGM",
        "sourceLayers": [
          {
            "_id": "vt0",
            "_rank": 1,
            "exclusiveGroup": "main",
            "followingItems": [],
            "name": "VT",
            "pieces": [
              {
                "instance": {
                  "_id": "randomId9002_part0_0_tmp_instance_randomId9002_piece001",
                  "isTemporary": true,
                  "partInstanceId": "randomId9002_part0_0_tmp_instance",
                  "piece": {
                    "_id": "randomId9002_piece001",
                    "content": {},
                    "enable": {
                      "start": 0,
                    },
                    "externalId": "MOCK_PIECE_001",
                    "invalid": false,
                    "lifespan": "part-only",
                    "name": "Piece 001",
                    "outputLayerId": "pgm",
                    "pieceType": "normal",
                    "sourceLayerId": "vt0",
                    "startPartId": "randomId9002_part0_0",
                    "startRundownId": "randomId9002",
                    "startSegmentId": "randomId9002_segment0",
                    "timelineObjectsString": "[]",
                  },
                  "playlistActivationId": "",
                  "priority": 5,
                  "rundownId": "randomId9002",
                },
                "outputLayer": [Circular],
                "renderedDuration": null,
                "renderedInPoint": 0,
                "sourceLayer": [Circular],
              },
            ],
            "type": 2,
          },
          {
            "_id": "cam0",
            "_rank": 0,
            "exclusiveGroup": "main",
            "followingItems": [],
            "name": "Camera",
            "pieces": [
              {
                "instance": {
                  "_id": "randomId9002_part0_1_tmp_instance_randomId9002_piece010",
                  "isTemporary": true,
                  "partInstanceId": "randomId9002_part0_1_tmp_instance",
                  "piece": {
                    "_id": "randomId9002_piece010",
                    "content": {},
                    "enable": {
                      "start": 0,
                    },
                    "externalId": "MOCK_PIECE_010",
                    "invalid": false,
                    "lifespan": "part-only",
                    "name": "Piece 010",
                    "outputLayerId": "pgm",
                    "pieceType": "normal",
                    "sourceLayerId": "cam0",
                    "startPartId": "randomId9002_part0_1",
                    "startRundownId": "randomId9002",
                    "startSegmentId": "randomId9002_segment0",
                    "timelineObjectsString": "[]",
                  },
                  "playlistActivationId": "",
                  "priority": 5,
                  "rundownId": "randomId9002",
                },
                "outputLayer": [Circular],
                "renderedDuration": null,
                "renderedInPoint": 0,
                "sourceLayer": [Circular],
              },
            ],
            "type": 1,
          },
        ],
        "used": true,
      },
    },
    "rundownId": "randomId9002",
    "sourceLayers": {
      "cam0": {
        "_id": "cam0",
        "_rank": 0,
        "exclusiveGroup": "main",
        "followingItems": [],
        "name": "Camera",
        "pieces": [],
        "type": 1,
      },
      "graphics0": {
        "_id": "graphics0",
        "_rank": 3,
        "followingItems": [],
        "name": "Graphic",
        "pieces": [],
        "type": 5,
      },
      "transition0": {
        "_id": "transition0",
        "_rank": 2,
        "followingItems": [],
        "name": "Transition",
        "pieces": [],
        "type": 13,
      },
      "vt0": {
        "_id": "vt0",
        "_rank": 1,
        "exclusiveGroup": "main",
        "followingItems": [],
        "name": "VT",
        "pieces": [],
        "type": 2,
      },
    },
  },
}
`;
